Step function: correct_bad_pixels

hsergi opened the pull request:
This PR covers the use of BadPixelMap to correct bad pixels in the data. UT goes over agreed sub-tests, including CR.

maxwellmb commented on corgidrp/l2a_to_l2b.py:
```diff
@@ -171,13 +171,33 @@ def correct_bad_pixels(input_dataset):
+        # load CR mask
+        cr_mask = dq_cube[i]
+        # combine CR and BP masks
+        bp_cr_mask = (cr_mask+bp_mask[0].data).astype(int)
```
Kevin brought up a good point in the thread for PR #93. Rather than just add things here, I think we want to break it into bits. I think in the long run it'll be safer and more robust.

maxwellmb replied:
I think this can be done pretty easily with `numpy.packbits`, `numpy.unpackbits` and the `numpy.bitwise_or`.

hsergi replied:
Sounds good. I understood that was a method that could be implemented later on. I'll implement it now

hsergi replied:
@maxwellmb: Main has dq as 0 or 1: https://github.com/roman-corgi/corgidrp/blob/de9f4b00c6b8ffe4c2a26bfd2bb80b4cf090209f/corgidrp/data.py#L172

I'm not sure now which data type is the numpy array in dq by looking at data.py if it's not up to date with the DRP implementation document. Looking at other tests, dq are integers. 

If they are integers, are yiu suggesting to cast them into bits arrays, checking bit array values and performing operations with bits, to come back to integers when dq is stored again?

maxwellmb replied:
> @maxwellmb: Main has dq as 0 or 1:
> 
> https://github.com/roman-corgi/corgidrp/blob/de9f4b00c6b8ffe4c2a26bfd2bb80b4cf090209f/corgidrp/data.py#L172

Yes, I think that line is out of date. Good catch! Do you mind updating it? 

> I'm not sure now which data type is the numpy array in dq by looking at data.py if it's not up to date with the DRP implementation document. Looking at other tests, dq are integers.

Yes, I _think_ the dq should be uint16

> 
> If they are integers, are you suggesting to cast them into bit arrays, checking bit array values and performing operations with bits, to come back to integers when dq is stored again?

Yup!

hsergi replied:
https://numpy.org/doc/stable/reference/generated/numpy.unpackbits.html#numpy.unpackbits requires np.uint8 to work. It does not if it is np.uint16. I'll be using uint8.

hsergi replied:
As for the doc string, we agreed to move it to a minor Issue: https://github.com/roman-corgi/corgidrp/issues/103

hsergi replied:
Done. I resubmitted both l2a_to_l2b.py and test_badpixel.

maxwellmb commented on tests/test_badpixel.py:
```diff
@@ -0,0 +1,91 @@
+
+def test_bad_pixels():
+
+    corgidrp.track_individual_errors = True
```
tracking individual errors is a best effort mode with the default mode being to not track the errors. I think we probably want to set this to False for testing

maxwellmb commented on corgidrp/l2a_to_l2b.py:
```diff
@@ -171,13 +171,33 @@ def correct_bad_pixels(input_dataset):
+        bp = np.where(bp_cr_mask != 0)
+        data_cube[i, bp[0], bp[1]] = np.nan
+        dq_cube[i] = bp_cr_mask
+
```
Right now in the implementation doc, we're using the 1st bit to signify that a pixel has been replaced. Although in this case, since it's a nan, it'll be obvious I think we still probably want to update that bit in the DQ too.

hsergi replied:
Not sure I follow. The image data pixels that are replaced by NaN are those that have a non-zero bit in their DQ. That is, they already have it. The updated DQ value for each pixel on the image data happens a step earlier, not in these lines.

maxwellmb replied:
The DQ has different bits that indicate different things. One of the bits (bit 1) indicates that a pixel's value has been replaced. Since we're replacing it here, then I think we should set that bit.

hsergi replied:
@maxwellmb: Is it not set already? 

`# combine DQ and BP masks`
  `bp_dq_mask_bit = np.bitwise_or(dq_mask_bit, bp_mask_bit)`
        `# Transform back to integer DQ values`
        `bp_dq_mask = np.packbits(bp_dq_mask_bit).reshape(dq_mask.shape[0], dq_mask.shape[1])`
       ` # mask affected pixels with NaN`
        `bp = np.where(bp_dq_mask != 0)`

I do not see why DQ values need to be set after the last line if they are already set to a non-good pixel value. That's what is used to replace a pixel on the image data by NaN. What would be the new setting for the DQ pixel? It should be the same non-zero value. DQ values are not bitarrays. They are integers, and you asked me to combine them as bitarrays for operations, only. I'm also confused as to what bit 1 means (first?, or second if we count first index as 0). Can you add an example of what you would do because I'm not getting it? Thanks a lot!

maxwellmb commented on tests/test_badpixel.py:
```diff
@@ -0,0 +1,91 @@
+    row_cr_test = [546, 789, 123, 43, 547, 675]
+    for i_col in col_cr_test:
+        for i_row in row_cr_test:
+            dq_cube[0, i_col, i_row] += 128
```
Similar to above, can we do this by setting the bit?

hsergi replied:
Done

hsergi replied:
UT passed

hsergi commented on corgidrp/l2a_to_l2b.py:
```diff
@@ -171,13 +171,33 @@ def correct_bad_pixels(input_dataset):
+
+    for i in range(data_cube.shape[0]):
+        # load CR mask
+        cr_mask = dq_cube[i]
```
I understood that at this level DQ has only CR. It's agnostic to the type of bad pixel, so it may be better to simply change the name and comments

hsergi commented:
As for the DQ doc string, I have: "dq (np.array): 2-D data quality, 0: good. Other values track different causes for bad pixels and other pixel-level effects in accordance with the DRP implementation document."

maxwellmb commented:
@hsergi, please resolve the conflicts and then I'll take a look at this.

hsergi commented:
@maxwellmb . I think I addressed all the comments. Outstanding ones: doc string for dq and using bitwise_or only. There was one more that I may misunderstand though bits are already set to DQ values before image data are replaced by NaN. I think it's fine as-is. If I am not getting the point, can you please tell me what it is missing exactly? The full comment is above.

hsergi commented:
@maxwellmb: I _think_ I have addressed your comment about setting the bit of "2" to 1 in both the correction function (https://github.com/roman-corgi/corgidrp/blob/c9a6385a1da8eedd2029ba02f5202ca36ea0516a/corgidrp/l2a_to_l2b.py#L195) and the test function (https://github.com/roman-corgi/corgidrp/blob/c9a6385a1da8eedd2029ba02f5202ca36ea0516a/tests/test_badpixel.py#L67 to https://github.com/roman-corgi/corgidrp/blob/c9a6385a1da8eedd2029ba02f5202ca36ea0516a/tests/test_badpixel.py#L79)

maxwellmb commented:
Alright, a few last minor thoughts: 
- You can delete the "MMB thoughts" now that things are implemented. 
- It would be great to make the docstring header a bit more descriptive (e.g. clarify how bad pixels are identified and what we do to replace the). 
- The "Returns" section says we're replacing by 0s, but at the moment its NaNs. 
- The description of the bp_mask input should instead say that this is a calibration file, not a bad pixel map for a specific frame. 


Also, there are still merge conflicts that you need to resolve. 

Otherwise we're good to go!
